### PR TITLE
perf: only compute start offset for overlong lines

### PR DESCRIPTION
**Pull request by sbrugman**
Moves the computation of the `start_offset` for overlong lines to just before the result is returned. There is a slight overhead for overlong lines (double the work for the first `limit` characters). 

In practice this results in a speedup on the CPython codebase. Most lines are not overlong, or are not enforced because the line ends with a URL, or does not contain whitespace. Nonetheless, the 0.3% of overlong lines are a lot compared to other violations.

### Before
selected before
_Selected W505 and E501_

all before
_All rules_

### After
selected after
_Selected W505 and E501_

all after
_All rules_

CPython line statistics:
- Number of Python lines: 867.696
- Number of overlong lines: 2.963 (0.3%)

<details>

Benchmark selected:
```shell
cargo build --release && hyperfine --warmup 10 --min-runs 50 \                                                  
  "./target/release/ruff ./crates/ruff/resources/test/cpython/ --no-cache -e --select W505,E501"
```

Benchmark all:
```shell
cargo build --release && hyperfine --warmup 10 --min-runs 50 \                                                  
  "./target/release/ruff ./crates/ruff/resources/test/cpython/ --no-cache -e --select ALL"
```

Overlong lines in CPython

```shell
cargo run -p ruff_cli -- check crates/ruff/resources/test/cpython/Lib --no-cache --select=E501,W505 --statistics
```

Total Python lines:
```shell
find crates/ruff/resources/test/cpython/ -name '*.py' | xargs wc -l
```

</details>

(Performance tested on Mac M1)

### CODE REVIEW

File: crates/ruff/src/rules/pycodestyle/helpers.rs [hunk 1]
```diff
@@ -91,6 +83,17 @@ pub(super) fn is_overlong(
         }
     }
 
+    // Obtain the start offset of the part of te line that exceeds the limit
```
Comment (konstin) [inline crates/ruff/src/rules/pycodestyle/helpers.rs:86]:
```suggestion
    // Obtain the start offset of the part of the line that exceeds the limit
```

**Review by konstin — APPROVED** [review]
amazing, that's a neat speedup

**Comment by charliermarsh** [comment]
That’s awesome. I’m wondering if there’s another optimization we can make here. Is it faster to compute the number of bytes in the line instead of the width? If so, could we only check lines that have more bytes than the line limit? I think it would be impossible for an overlong line to have _fewer_ bytes than the limit, though lines with more bytes could end up being below the limit.

**Comment by MichaReiser** [comment]
> That’s awesome. I’m wondering if there’s another optimization we can make here. Is it faster to compute the number of bytes in the line instead of the width? If so, could we only check lines that have more bytes than the line limit? I think it would be impossible for an overlong line to have _fewer_ bytes than the limit, though lines with more bytes could end up being below the limit.

That sounds like a neat optimization. Each character is between 1-4 bytes. Meaning your assumption here is safe. 

Checking the byte length is `O(1)`

**Comment by sbrugman** [comment]
![check byte length(https://github.com/astral-sh/ruff/assets/9756388/0b262119-ab1e-4aaf-9315-be2d6f170f18)

**Pull request merged**